File: app/src/main/java/com/example/taras/reminerapp/data/RemindViewModel.java
package com.example.taras.reminerapp.data;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.support.annotation.NonNull;

import com.example.taras.reminerapp.App;
import com.example.taras.reminerapp.db.AppDatabase;
import com.example.taras.reminerapp.db.Constants;
import com.example.taras.reminerapp.db.model.Remind;

import java.util.List;

/**
 * Created by Taras Koloshmatin on 30.07.2018
 */
public class RemindViewModel extends ViewModel {

    private final LiveData<List<Remind>> mRemindList;
    private final LiveData<List<Remind>> mUserRemindList;
    private final LiveData<List<Remind>> mNewsList;
    private final LiveData<List<Remind>> mEventList;
    private final LiveData<List<Remind>> mVideoList;

    public RemindViewModel() {
        mRemindList = AppDatabase.getInstance().remindDao().getList();
        mUserRemindList = AppDatabase.getInstance().remindDao().getListByType(Constants.TYPE_USER_REMIND);
        mNewsList = AppDatabase.getInstance().remindDao().getListByType(Constants.TYPE_NEWS);
        mEventList = AppDatabase.getInstance().remindDao().getListByType(Constants.TYPE_EVENT);
        mVideoList = AppDatabase.getInstance().remindDao().getListByType(Constants.TYPE_VIDEO);
    }


    public LiveData<List<Remind>> getRemindList() {
        return mRemindList;
    }

    public LiveData<List<Remind>> getUserRemindList() {
        return mUserRemindList;
    }

    public LiveData<List<Remind>> getNewsList() {
        return mNewsList;
    }

    public LiveData<List<Remind>> getEventList() {
        return mEventList;
    }

    public LiveData<List<Remind>> getVideoList() {
        return mVideoList;
    }

//    public void deleteItem(Remind remind) {
//        new deleteAsyncTask().execute(remind);
//    }
//
//    private static class deleteAsyncTask extends AsyncTask<Remind, Void, Void> {
//        @Override
//        protected Void doInBackground(final Remind... params) {
//            AppDatabase.getInstance().remindDao().delete(params[0]);
//            return null;
//        }
//    }

    public static class RemindViewModelFactory extends ViewModelProvider.NewInstanceFactory {

        @NonNull
        private final App application;

        public RemindViewModelFactory(@NonNull App application) {
            this.application = application;
        }

        @NonNull
        @Override
        public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
            return (T) new RemindViewModel();
        }
    }
}

File: app/src/main/java/com/example/taras/reminerapp/data/RemindViewModel.kt
package com.example.taras.reminerapp.data

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.MutableLiveData
import android.arch.lifecycle.ViewModel
import android.arch.lifecycle.ViewModelProvider
import android.os.Bundle
import com.example.taras.reminerapp.db.AppDatabase
import com.example.taras.reminerapp.db.Constants
import com.example.taras.reminerapp.db.model.Remind
import org.jetbrains.anko.doAsync

/**
 * Created by Taras Koloshmatin on 31.07.2018
 */
class RemindViewModel : ViewModel() {

    private val KEY = "key"

    private var remind = Remind("TITLE", "20.20.20", "DESCRIPTION", "type_view_model")

    private var mRemindList: LiveData<List<Remind>> = AppDatabase.getInstance().remindDao().getList()
    private var mUserRemindList: LiveData<List<Remind>> = AppDatabase.getInstance().remindDao().getListByType(Constants.TYPE_USER_REMIND)
    private var mNewsList: LiveData<List<Remind>> = AppDatabase.getInstance().remindDao().getListByType(Constants.TYPE_NEWS)
    private var mEventList: LiveData<List<Remind>> = AppDatabase.getInstance().remindDao().getListByType(Constants.TYPE_EVENT)
    private var mVideoList: LiveData<List<Remind>> = AppDatabase.getInstance().remindDao().getListByType(Constants.TYPE_VIDEO)

    private var mStarList: MutableLiveData<List<Remind>> = MutableLiveData()

    fun getRemindList(): LiveData<List<Remind>> {
        return mRemindList
    }

    fun getUserRemindList(): LiveData<List<Remind>> {
        return mUserRemindList
    }

    fun getNewsList(): LiveData<List<Remind>> {
        return mNewsList
    }

    fun getEventList(): LiveData<List<Remind>> {
        return mEventList
    }

    fun getVideoList(): LiveData<List<Remind>> {
        return mVideoList
    }


    fun saveState(outState: Bundle) {
        outState.putParcelable(KEY, remind)
    }

    fun restoreState(bundle: Bundle?) {
        bundle?.let {
            remind = bundle.getParcelable(KEY)
            doAsync {
                AppDatabase.getInstance().remindDao().insert(remind)
            }
        }
    }


    fun setStar(remind: Remind) {
        var list: List<Remind> = listOf(remind)
        mStarList.value = list
    }

    fun getStarList(): LiveData<List<Remind>> {
        return mStarList
    }


    class Factory : ViewModelProvider.NewInstanceFactory() {
        override fun <T : ViewModel> create(modelClass: Class<T>): T {
            return RemindViewModel() as T
        }
    }
    /*
        usage:
        mBinding.model = ViewModelProviders.of(this@NewsFragment, Factory()).get(RemindViewModel::class.java)

     */


    // for different ViewModels:
    protected inline fun <VM : ViewModel> viewModelFactory(crossinline f: () -> VM) = object : ViewModelProvider.Factory {
        override fun <T : ViewModel> create(aClass: Class<T>): T = f() as T
    }
    /*
        usage:
        mBinding.model = ViewModelProviders.of(this@NewsFragment, viewModelFactory { ReminderViewModel() }).get(RemindViewModel::class.java)
     */
}